src/Native/System.IO.Compression.Native/pal_zlib.cpp

```diff
@@ -0,0 +1,184 @@
+
+static_assert(PAL_Z_NOFLUSH == Z_NO_FLUSH, "");
+static_assert(PAL_Z_PARTIALFLUSH == Z_PARTIAL_FLUSH, "");
+static_assert(PAL_Z_SYNCFLUSH == Z_SYNC_FLUSH, "");
+static_assert(PAL_Z_FULLFLUSH == Z_FULL_FLUSH, "");
+static_assert(PAL_Z_FINISH == Z_FINISH, "");
+static_assert(PAL_Z_BLOCK == Z_BLOCK, "");
+
+static_assert(PAL_Z_OK == Z_OK, "");
+static_assert(PAL_Z_STREAMEND == Z_STREAM_END, "");
+static_assert(PAL_Z_NEEDDICTIONARY == Z_NEED_DICT, "");
+static_assert(PAL_Z_ERRORNO == Z_ERRNO, "");
+static_assert(PAL_Z_STREAMERROR == Z_STREAM_ERROR, "");
+static_assert(PAL_Z_DATAERROR == Z_DATA_ERROR, "");
+static_assert(PAL_Z_MEMERROR == Z_MEM_ERROR, "");
+static_assert(PAL_Z_BUFERROR == Z_BUF_ERROR, "");
+static_assert(PAL_Z_VERSIONERROR == Z_VERSION_ERROR, "");
+
+static_assert(PAL_Z_NOCOMPRESSION == Z_NO_COMPRESSION, "");
+static_assert(PAL_Z_BESTSPEED == Z_BEST_SPEED, "");
+static_assert(PAL_Z_BESTCOMPRESSION == Z_BEST_COMPRESSION, "");
+static_assert(PAL_Z_DEFAULTCOMPRESSION == Z_DEFAULT_COMPRESSION, "");
+
+static_assert(PAL_Z_FILTERED == Z_FILTERED, "");
+static_assert(PAL_Z_HUFFMANONLY == Z_HUFFMAN_ONLY, "");
+static_assert(PAL_Z_RLE == Z_RLE, "");
+static_assert(PAL_Z_FIXED == Z_FIXED, "");
+static_assert(PAL_Z_DEFAULTSTRATEGY == Z_DEFAULT_STRATEGY, "");
+
+static_assert(PAL_Z_DEFLATED == Z_DEFLATED, "");
+
+/*
+Initializes the PAL_ZStream by creating and setting its underlying z_stream.
+*/
+static void Init(PAL_ZStream* stream)
+{
+    z_stream* zStream = new z_stream();
+    zStream->zalloc = Z_NULL;
+    zStream->zfree = Z_NULL;
+    zStream->opaque = Z_NULL;
+
+    stream->internalState = zStream;
+}
+
+/*
+Frees any memory on the PAL_ZStream that was created by Init.
+*/
+static void End(PAL_ZStream* stream)
+{
+    z_stream* zStream = reinterpret_cast<z_stream*>(stream->internalState);
+
+    delete zStream;
+    stream->internalState = nullptr;
+}
+
+/*
+Transfers the output values from the underlying z_stream to the PAL_ZStream.
+*/
+static void TransferState(z_stream* from, PAL_ZStream* to)
+{
+    to->nextIn = from->next_in;
+    to->availIn = from->avail_in;
+
+    to->nextOut = from->next_out;
+    to->availOut = from->avail_out;
+
+    to->msg = from->msg;
+}
+
+/*
+Transfers the input values from the PAL_ZStream to the underlying z_stream object.
+*/
+static void TransferState(PAL_ZStream* from, z_stream* to)
+{
+    to->next_in = from->nextIn;
+    to->avail_in = from->availIn;
+
+    to->next_out = from->nextOut;
+    to->avail_out = from->availOut;
+}
+
+/*
+Gets the current z_stream object for the specified PAL_ZStream.
+
+This ensures any inputs are transferred from the PAL_ZStream to the underlying z_stream,
+since the current values are always needed.
+*/
+static z_stream* GetCurrentZStream(PAL_ZStream* stream)
+{
+    z_stream* zStream = reinterpret_cast<z_stream*>(stream->internalState);
+    TransferState(stream, zStream);
+    return zStream;
+}
+
+extern "C" int32_t IsZLibAvailable()
+{
+    return true;
+}
```

Comment:
I really struggled with this. So I decided to make it simple for now and have a conversation about what to do here.

My thinking was that if zlib isn't available on the platform, then we could do a #if HAVE_ZLIB return true; #else return false; here.  However, all the platforms we are supporting: Ubuntu, OSX, and CentOS all have it available. I'm not sure on FreeBSD yet.
So I was thinking to "rough the support in" like this.

If I was to take the shim IsZLibAvailable() out and return `WorkerType.ZLib` in DeflateStream.Unix.cs, then both Windows and Unix would always use the ZLib in/deflators, which means the Managed backup code would be dead code. So should that then be removed? I didn't want to go that far in this initial implementation.

I figured if we still wanted to support not having a zlib installed on the platform, this seemed like the most appropriate approach to enable that - a compile time check that would make this method return false. But since I don't know of a platform that doesn't have zlib, I wouldn't be able to test the compile time check, so I didn't add it (yet).

Another option would be to do a LoadLibrary call here instead.

So I think our first question to answer is: as of now, do we still want to have support for Unix systems that don't have zlib installed?